src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestIT.java

```diff
@@ -1099,86 +1099,181 @@ public void testDataTypes() {
 
     SnowflakeStreamingIngestChannel channel = openChannel(tableName, "CHANNEL");
 
-    Random r = new Random();
-    for (int val = 0; val < 10000; val++) {
-      verifyInsertValidationResponse(channel.insertRow(getRandomRow(r), Integer.toString(val)));
-    }
-    waitChannelFlushed(channel, 10000);
-  }
-
-  private static Map<String, Object> getRandomRow(Random r) {
-    Map<String, Object> row = new HashMap<>();
-    row.put("num", r.nextInt());
-    row.put("num_10_5", nextFloat(r));
-    row.put(
-        "num_38_0",
-        new BigDecimal("" + nextLongOfPrecision(r, 18) + Math.abs(nextLongOfPrecision(r, 18))));
-    row.put("num_2_0", r.nextInt(100));
-    row.put("num_4_0", r.nextInt(10000));
-    row.put("num_9_0", r.nextInt(1000000000));
-    row.put("num_18_0", nextLongOfPrecision(r, 18));
-    row.put("num_float", nextFloat(r));
-    row.put("str_varchar", nextString(r));
-    row.put("bin", nextBytes(r));
-    row.put("bl", r.nextBoolean());
-    row.put("var", nextJson(r));
-    row.put("obj", nextJson(r));
-    row.put("arr", Arrays.asList(r.nextInt(100), r.nextInt(100), r.nextInt(100)));
-    row.put("epochdays", Math.abs(r.nextInt()) % 18963); // DATE, 02.12.2021
-    row.put(
-        "timesec",
-        (r.nextInt(11) * 60 * 60 + r.nextInt(59) * 60 + r.nextInt(59)) * 10000
-            + r.nextInt(9999)); // TIME(4), 05:12:43.4536
-    row.put(
-        "timenano",
-        (14 * 60 * 60 + 26 * 60 + 34) * 1000000000L + 437582643); // TIME(9), 14:26:34.437582643
-    row.put(
-        "epochsec", Math.abs(r.nextLong()) % 1638459438); // TIMESTAMP_LTZ(0), 02.12.2021 15:37:18
-    row.put(
+    Map<String, Object> negRow = new HashMap<>();
+    negRow.put("num", -123);
+    negRow.put("num_10_5", new BigDecimal("-12345.54321"));
+    negRow.put("num_38_0", new BigDecimal("-91234567899876543219876543211234567891"));
+    negRow.put("num_2_0", -12);
+    negRow.put("num_4_0", -1234);
+    negRow.put("num_9_0", -123456789);
+    negRow.put("num_18_0", -123456789987654321L);
+    negRow.put("num_float", -12345.54321f);
+    negRow.put("str_varchar", "zxccvbnm,.");
+    negRow.put("bin", new byte[] {1, 2, 3});
+    negRow.put("bl", true);
+    negRow.put("var", "{}");
+    negRow.put("obj", "{}");
+    negRow.put("arr", Arrays.asList());
+    negRow.put("epochdays", 0);
+    negRow.put("epochsec", 0);
+    negRow.put("epochnano", new BigDecimal("0"));
+    negRow.put("timesec", 0);
+    negRow.put("timenano", 0);
+    verifyInsertValidationResponse(channel.insertRow(negRow, Integer.toString(0)));
```

Comment:
This is a lot of details in one method.
I suggest to move them into separate functions `getNegRow()` etc, also assertion code for each row.

Reply:
ack.